apis/fluentd/v1alpha1/fluentd_types.go

```diff
@@ -41,9 +41,9 @@ type FluentdSpec struct {
 	// Select cluster output plugins used to send all logs that did not match any route to the matching outputs
 	DefaultOutputSelector *metav1.LabelSelector `json:"defaultOutputSelector,omitempty"`
 	// By default will build the related service according to the globalinputs definition.
-	DisableService bool `json:"disableService,omitempty"`
-	// Numbers of the Fluentd instance
-	Replicas *int32 `json:"replicas,omitempty"`
+	DisableService  bool            `json:"disableService,omitempty"`
+	DaemonSetSpec   DaemonSetSpec   `json:"daemonSetSpec,omitempty"`
```

Comment:
I suggest keeping all the original fields as they were as before while adding a switch like `mode: agent` and `mode: collector` to control whether it should be deployed as DaemonSet or StatefulSet.:
- Default to `mode: agent`
- Ignore StatefulSet specific fields when creating DaemonSet and vice versa

Reply:
Incorporated these changes. Although it would be confusing for customer to see replicas in values.yaml. As it is not generic property for fluentd. It is specific for the Fluend as statefulset...
Let me know, what do you think with the latest changes.

Reply:
> Although it would be confusing for customer to see replicas in values.yaml. As it is not generic property for fluentd. It is specific for the Fluend as statefulset...

@karan56625 Thanks for the changes. I think we can add more comments to make this more clear:

https://github.com/fluent/fluent-operator/pull/839/files#diff-411f99718ce3f5d20179ea69352728e41f32d06106e88d09b1f3ace86f22c8e4R45
```go
	// Numbers of the Fluentd instance
        // Applicable when the mode is "collector", and will be ignored when the mode is "agent"
	Replicas *int32 `json:"replicas,omitempty"`
``

https://github.com/fluent/fluent-operator/pull/839/files#diff-411f99718ce3f5d20179ea69352728e41f32d06106e88d09b1f3ace86f22c8e4R94
```go
	// volumeClaimTemplates is a list of claims that pods are allowed to reference.
	// The StatefulSet controller is responsible for mapping network identities to
	// claims in a way that maintains the identity of a pod. Every claim in
	// this list must have at least one matching (by name) volumeMount in one
	// container in the template.
        // Applicable when the mode is "collector", and will be ignored when the mode is "agent"
	VolumeClaimTemplates []corev1.PersistentVolumeClaim `json:"volumeClaimTemplates,omitempty"`
```

https://github.com/fluent/fluent-operator/pull/839/files#diff-411f99718ce3f5d20179ea69352728e41f32d06106e88d09b1f3ace86f22c8e4R108
```go
	// Storage for position db. You will use it if tail input is enabled.
        // Applicable when the mode is "agent", and will be ignored when the mode is "collector"
	PositionDB corev1.VolumeSource `json:"positionDB,omitempty"`
```

https://github.com/fluent/fluent-operator/pull/839/files#diff-c8382299239a550c6747ba79bff00ed81780ea6a5639d69cf192c0ea3a269bc6R276
```
        // Applicable when the mode is "collector", and will be ignored when the mode is "agent"
	replicas: 1
``

Reply:
updated